src/graph/immutable_graph.cc

```diff
@@ -144,328 +97,112 @@ CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids, bool is_multigraph,
   CHECK_EQ(indices->shape[0], edge_ids->shape[0]);
   const int64_t num_verts = indptr->shape[0] - 1;
   const int64_t num_edges = indices->shape[0];
-  std::tie(indptr_, indices_, edge_ids_) = MapFromSharedMemory(
+  adj_.num_rows = num_verts;
+  adj_.num_cols = num_verts;
+  std::tie(adj_.indptr, adj_.indices, adj_.data) = MapFromSharedMemory(
       shared_mem_name, num_verts, num_edges, true);
   // copy the given data into the shared memory arrays
-  indptr_.CopyFrom(indptr);
-  indices_.CopyFrom(indices);
-  edge_ids_.CopyFrom(edge_ids);
+  adj_.indptr.CopyFrom(indptr);
+  adj_.indices.CopyFrom(indices);
+  adj_.data.CopyFrom(edge_ids);
 }
 
 CSR::CSR(const std::string &shared_mem_name,
          int64_t num_verts, int64_t num_edges, bool is_multigraph)
   : is_multigraph_(is_multigraph), shared_mem_name_(shared_mem_name) {
-  std::tie(indptr_, indices_, edge_ids_) = MapFromSharedMemory(
+  CHECK(!(num_verts == 0 && num_edges != 0));
+  adj_.num_rows = num_verts;
+  adj_.num_cols = num_verts;
+  std::tie(adj_.indptr, adj_.indices, adj_.data) = MapFromSharedMemory(
       shared_mem_name, num_verts, num_edges, false);
 }
 
```

Comment:
Seems we are missing the implementation of 

```c++
  CSR(int64_t num_vertices, int64_t num_edges,
      IndptrIter indptr_begin, IndicesIter indices_begin, EdgeIdIter edge_ids_begin,
      bool is_multigraph);
```

declared in L44 of immutable_graph.h.